### PR TITLE
[4.0] Multilingual: Create Article using specific category improvement

### PR DESCRIPTION
**Pull request by infograf768**
### Summary of Changes
Multilingual site
When using a Create Article menu item using a specific category, whatever the language set for that category, the default language field is always set to the site language and can also be modified, which can cause unwanted errors. 
In a multilingual site, an article tagged to a specific language should only be in a category tagged to the same language or All languages.

1. The first correction comes from https://github.com/joomla/joomla-cms/pull/6966 where the intention was, when creating an article in **backend** to propose in the create article form by default the same language as the language filter articles manager was set to . Thanks to @Bakual for helping me finding this :)
As this was not limited to Administration, it was forcing as default the site default language when creating an article in **frontend**.

https://github.com/joomla/joomla-cms/blob/52fc4b86a8cb83fdae095f52977de3139e115522/administrator/components/com_content/src/Model/ArticleModel.php#L634

This is now corrected here by limiting it to admin

We never remarked that overseen aspect because we had already a way in frontend to do that in the HtmlView.php

https://github.com/joomla/joomla-cms/blob/52fc4b86a8cb83fdae095f52977de3139e115522/components/com_content/src/View/Form/HtmlView.php#L166-L171

2. To prevent a user a to submit an article with the wrong language when a specific category is selected in the menu item, we then first have to check if this parameter is set in the view.
If it is, then we should not propose as default the site language (which can be modifed) but instead code the FormModel `preprocessForm()` method to check the language of the category and, if not set to All languages, force the article to be set to the same language and read-only.

Note: The patch not only prevents error when the Create Article menu item deals with a specific category set to the same language as the site language, but also, cherry on the cake, let's create such menu items with a category set to another language than the site language. Both read-only.

For example:
One can have a Create Article menu item tagged to French and thus displayed in French frontend, with an English tagged specific category. The language field will be rightfully set as English per default and read-only.
As when one can do the more, one can do the less, let's test it.

### Testing Instructions
Multilingual site with 2 languages.
Let's do the cherry on the cake :)
Create a Create Article menu item tagged to French with a specific category tagged to English in a French tagged menu.
Set access to Special.
Login frontend and click on the menu item.
Check the category and the language before submitting.
Check the Articles Manager for the new article.

### Expected result AFTER applying this Pull Request
**Creating the menu item**
Screen Shot 2021-02-19 at 09 58 48

Screen Shot 2021-02-19 at 09 59 08

**Frontend menu item displayed**

Screen Shot 2021-02-19 at 09 59 45

**Article form showing category**

Screen Shot 2021-02-19 at 10 00 13

**Article form showing read-only language field**

Screen Shot 2021-02-19 at 10 00 25

**Articles Manager showing newly created article**

Screen Shot 2021-02-19 at 10 04 22

### Note
One can have multiple Create Article menu items set to use different specific category (See my screenshot above).
Beware: once there is (are) such Create Article menu item(s), there should never be also a Create Article menu item NOT set to use a specific category.

### CODE REVIEW

**Comment by richard67** [comment]
Does this PR also solve or answer https://github.com/joomla/joomla-cms/discussions/32369 ?

**Comment by infograf768** [comment]
> Does this PR also solve or answer #32369 ?

Only partly. That poster is very much confused. ;)
I replied to him.

File: components/com_content/src/Model/FormModel.php [hunk 1]
```diff
@@ -220,6 +220,24 @@ protected function preprocessForm(Form $form, $data, $group = 'content')
...
 			$form->setFieldAttribute('catid', 'readonly', 'true');
+
+			if (Multilanguage::isEnabled())
+			{
+				$db    = $this->getDbo();
+				$query = $db->getQuery(true)
+					->select($db->quoteName('language'))
+					->from($db->quoteName('#__categories'))
+					->where($db->quoteName('id') . ' = ' . $db->quote($params->get('catid')));
+				$db->setQuery($query);
```
Comment (Quy) [inline components/com_content/src/Model/FormModel.php]:
```suggestion
				$categoryId = (int) $params->get('catid');

				$db    = $this->getDbo();
				$query = $db->getQuery(true)
					->select($db->quoteName('language'))
					->from($db->quoteName('#__categories'))
					->where($db->quoteName('id') . ' = :categoryId')
					->bind(':categoryId', $categoryId, ParameterType::INTEGER);
				$db->setQuery($query);
```

Comment (Quy) [reply components/com_content/src/Model/FormModel.php]: Please add the following at the top:

`use Joomla\Database\ParameterType;`

**Comment by infograf768** [comment]
@Quy Done. Tested changes OK without really knowing why it is better. ;)

**Comment by Razzo1987** [comment]
I have tested this item :white_check_mark: successfully on 9434ce65a708c53d941a7eaa1f0cf51a065fce52<br />
Test FR 1 before patch
Test FR 2 after patch
immagine

<br />Work correctly<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32462">issues.joomla.org/tracker/joomla-cms/32462</a>.</sub>

**Comment by Quy** [comment]
> @Quy Done. Tested changes OK without really knowing why it is better. ;)

To reduce the exposure of a code base to SQL Injection attacks by sending the query and the data separately.

For more information: https://www.techfry.com/joomla/prepared-statements-in-joomla-4

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 5153193dca7cd4ed4a948ea23e6fed7b786c1002<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32462">issues.joomla.org/tracker/joomla-cms/32462</a>.</sub>

**Comment by infograf768** [comment]
rtc

tks for testing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32462">issues.joomla.org/tracker/joomla-cms/32462</a>.</sub>

**Comment by richard67** [comment]
The commit after @Razzo1987 's test was just a clean branch update, so that test is still valid. I've restored the test result in the issue tracker so it's properly counted.

**Comment by chmst** [comment]
Thanks!

**Comment by ghost** [comment]
@infograf768  Can you explain how to test this One can have multiple Create Article menu items set to use different specific category ?

What is expected result if you create 10 article menu items set to use different specific category ?

**Comment by drmenzelit** [comment]
@chnnst you can create several menu items of type "create article" and in each one you define the category where the article should be stored.
Menu 1: Create article for Blog (Category Blog selected)
Menu 2: Create article for News (Category News selected)
Menu 3: Create article for Politics (Category Politics selected)
.....

**Pull request merged**